draft-ietf-mls-protocol.md

```diff
@@ -1444,110 +1452,169 @@ To authenticate a message from a particular member, signatures are required.
 
-All the traffic keying material is recomputed whenever the underlying
-Application Secret changes.
+~~~~
+struct {
+  opaque gshash<0..2^32-1> = Hash(GroupState_[n]);
+  opaque leaf_index<0..2^16-1>;
+  uint32 ratchet_position;
+} RatchetContext
+~~~~
+
+The identity field is copied from the eponymous field in the BasicCredential
+struct. Further, ratchet_position is initialized to 1 for the first
+HKDF-Expand-Label call of a new ratchet and incremented for each subsequent call
+in that ratchet.
+
+HashRatCont[i, j] denotes the hash of a RatchetContext variable where:
+
+- gshash = Hash(GropuState_[n])
+
+- identity = i
+
+- ratchet_position = j
+
+Using this notation the ratchet for the group member at leaf i is defined to be:
+
+~~~~
+application_[i]_[j]_secret
+      |
+      +--> HKDF-Expand-Label(., "app-nonce", HashRatCont[i,j+1], AEAD.nonceLen)
+      |    = application_[i]_[j+1]_nonce
+      |
+      +--> HKDF-Expand-Label(., "app-key", HashRatCont[i,j+1], AEAD.keyLength)
+      |    = application_[i]_[j+1]_key
+      |
+      V
+HKDF-Expand-Label(., "app-secret", HashRatCont[i,j+1], Hash.length)
+= application_[i]_[j+1]_secret
+~~~~
+
+Here, AEAD.nonceLen and AEAD.keyLength denote the lengths in bytes of the
+nonce and key for the AEAD scheme defined by the ciphersuite.
+
+## Deletion Schedule {#deletion-schedule}
+
+It is important to delete all security sensitive values as soon as they, or
+another value derived from them, is used for encryption or decryption.
+
+More precisely, the values application_[i]\_[j]\_key and
+application\_[i]\_[j]_nonce are said to be "consumed" if they were used either
+to:
+
+* encrypt and send a message or
+* to successfully decrypt and authenticate a received message.
+
+More generally, a secret S is "consumed" if any key, nonce or secret derived
+from S has been consumed. (This goes both for values derived via Derive-Secret
+and HKDF-Expand-Label). Here, S may be the init_secret, update_secret,
+epoch_secret, application_secret as well as any secret in the AS Tree or one of
+the ratchets.
+
+As soon as a group member consumes a key, nonce or secret they MUST immediately
+delete (all representations of) that value. This is crucial to ensuring Forward
+Secrecy for past messages. Members MAY keep unconsumed secrets, keys and nonces
+around for some reasonable amount of time even if their generating secret was
+already consumed (e.g. due to out of order message delivery).
+
+For example, suppose a group member encrypts or successfully decrypts a message
+using the j-th key and nonce in the i-th ratchet. Then, for that member, all of
+the following values (and possibly more) have been consumed and MUST be deleted:
+
+* the init_secret, update_secret, epoch_secret, application_secret of that
+epoch,
+* all node secrets in the AS Tree on the path from the root to the leaf with
+index i,
+* the first j secrets in the i-th ratchet and
+* application_[i]\_[j]\_key and application_[i]\_[j]\_nonce.
```

Comment:
A diagram might help here.  E.g., in the following scenario...

```
      G
    /   \
   /     \
  E       F
 / \     / \
A0  B0  C0  D0 -+- KD0
            |   |
            |   +- ND0
            |
            D1 -+- KD1
            |   |
            |   +- ND1
            |
            D2 -+- KD2
                |
                +- ND2
```

When KD1/ND1 are used to decrypt a message, the following are:
* Consumed: G, F, D0, D1, KD1, ND1
* Not necessarily consumed: E, A0, B0, C0, KD0, ND0, D2, KD2, ND2

Reply:
Deriving a value doesnt mean the parent needs to be consumed. Only when the derived values are actually used/consumed does the parent need to be consumed. My understanding of this example is that only KD1/ND1 are used but not KD2/ND2. So D2 doesn't need to be consumed yet.

Reply:
Again, if D2 gets leaked, it strictly worse than leaking D3 because you get KD2 and ND2 as well...